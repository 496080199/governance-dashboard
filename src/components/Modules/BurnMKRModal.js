import React, { Fragment, useState } from 'react';
import { createPortal } from 'react-dom';
import styled from 'styled-components';
import { Button } from '@makerdao/ui-components-core';

const ModalCover = styled.aside`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 1000;
  transform: translateZ(0);
  background-color: rgba(72, 73, 95, 0.3);
`;

const Modal = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding: 2.5em 1.5em 1.5em 1.5em;
  background-color: #ffffff;
  border-radius: 4px;
  overflow-y: auto;
  -webkit-overflow-scrolling: touch;

  @media screen and (min-width: 500px) {
    left: 50%;
    top: 50%;
    height: auto;
    transform: translate(-50%, -50%);
    max-width: 30em;
    max-height: calc(100% - 1em);
  }
`;

const ModalClose = styled.button`
  position: absolute;
  top: 0;
  right: 0;
  padding: 0.5em;
  line-height: 1;
  border: 0;
  box-shadow: 0;
  cursor: pointer;
`;

const ModalCloseIcon = styled.svg`
  width: 21px;
  height: 21px;
  fill: rgba(#000, 0);
  stroke: black;
  stroke-linecap: round;
  stroke-width: 2;
`;

const ScreenReaderText = styled.span`
  border: 0 !important;
  clip: rect(0 0 0 0) !important;
  height: 1px !important;
  margin: -1px !important;
  overflow: hidden !important;
  padding: 0 !important;
  position: absolute !important;
  width: 1px !important;
  white-space: nowrap !important;
`;

const ModalBody = styled.div`
  padding-top: 0.25em;
`;

const ModalTrigger = ({ text, onOpen }) => (
  <Button variant="danger-outline" onClick={onOpen}>
    {text}
  </Button>
);

const ModalContent = ({ content, onClose }) => {
  return createPortal(
    <ModalCover>
      <Modal>
        <ModalClose onClick={onClose}>
          <ScreenReaderText>Close</ScreenReaderText>
          <ModalCloseIcon viewBox="0 0 40 40">
            <path d="M 10,10 L 30,30 M 30,10 L 10,30"></path>
          </ModalCloseIcon>
        </ModalClose>
        <ModalBody></ModalBody>
      </Modal>
    </ModalCover>,
    document.body
  );
};

export default props => {
  const { triggerText } = props;
  const [isOpen, setIsOpen] = useState(false);
  const onOpen = () => setIsOpen(true);
  const onClose = () => setIsOpen(false);
  return (
    <Fragment>
      <ModalTrigger text={triggerText} onOpen={onOpen} />
      {isOpen && <ModalContent onClose={onClose} />}
    </Fragment>
  );
};
