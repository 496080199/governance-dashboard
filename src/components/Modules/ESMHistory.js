import React, { useState } from 'react';
import { Card, Grid, Text } from '@makerdao/ui-components-core';

export default () => {
  const [stakingHistory, setStakingHistory] = useState(false);
  return (
    <Grid gridRowGap="m" my={'s'}>
      <Text.h4 textAlign="left" fontWeight="700">
        ESM History
      </Text.h4>
      <Card>
        {!stakingHistory ? (
          <Text.p
            color="#9FAFB9"
            fontWeight="300"
            alignSelf="center"
            p={'xl'}
            textAlign="center"
          >
            No history to show
          </Text.p>
        ) : null}
      </Card>
    </Grid>
  );
};
