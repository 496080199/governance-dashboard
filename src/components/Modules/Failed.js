import React from 'react';
import { netIdToName } from '../../utils/ethereum';
import { Grid, Text, Button, Link } from '@makerdao/ui-components-core';
import arrowTopRight from '../../imgs/arrowTopRight.svg';
import { etherscanLink } from '../../utils/ui';

function Failed(props) {
  const { onClose, txHash } = props;

  return (
    <Grid gridRowGap="m" mx={'s'}>
      <Text.h2 textAlign="center">{props.title}</Text.h2>
      <Text.p fontSize="1.7rem" color="darkLavender" textAlign="center">
        {props.subtitle}
      </Text.p>
      {txHash ? (
        <Link
          justifySelf="center"
          target="_blank"
          href={etherscanLink(
            txHash,
            netIdToName(window.maker.service('web3').networkId())
          )}
        >
          <Button
            my="xs"
            justifySelf="center"
            fontSize="s"
            py="xs"
            px="s"
            variant="secondary"
          >
            View transaction details{' '}
            <img alt="link arrow" src={arrowTopRight} />
          </Button>
        </Link>
      ) : (
        ''
      )}
      <Grid gridRowGap="s" justifySelf="center">
        <Button variant="secondary-outline" onClick={onClose}>
          Exit
        </Button>
      </Grid>
    </Grid>
  );
}
export default Failed;
