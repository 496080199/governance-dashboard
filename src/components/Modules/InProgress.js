import React, { Fragment, useState, useEffect } from 'react';
import { netIdToName } from '../../utils/ethereum';
import { Button, Text, Link } from '@makerdao/ui-components-core';
import arrowTopRight from '../../imgs/arrowTopRight.svg';
import LoadingBar from '../LoadingBar';
import { etherscanLink } from '../../utils/ui';

const InProgress = ({ onClose, txHash, title }) => {
  const [waitTime, setWaitTime] = useState();
  const maker = window.maker;
  useEffect(() => {
    (async () => {
      // this is the default transaction speed
      const waitTime = await maker.service('gas').getWaitTime('fast');
      const minutes = Math.round(waitTime);
      const seconds = Math.round(waitTime * 6) * 10;

      const waitTimeText =
        waitTime < 1
          ? `${seconds} seconds`
          : `${minutes} minute${minutes === 1 ? '' : 's'}`;

      setWaitTime(waitTimeText);
    })();
  }, [maker]);

  return (
    <Fragment>
      <Text.h2 mt="m" textAlign="center">
        {title}
      </Text.h2>
      <Text
        style={{ fontSize: 17, color: '#48495F' }}
        mt="m"
        textAlign="center"
      >
        The estimated time is {waitTime || 'being calculated'}. You can close
        this modal.
      </Text>
      {txHash && (
        <Link
          justifySelf="center"
          target="_blank"
          mt="m"
          href={etherscanLink(
            txHash,
            netIdToName(window.maker.service('web3').networkId())
          )}
        >
          <Button
            justifySelf="center"
            fontSize="s"
            py="xs"
            px="s"
            variant="secondary"
          >
            View transaction details{' '}
            <img alt="top right arrow" src={arrowTopRight} />
          </Button>
        </Link>
      )}
      <LoadingBar />
      <Button
        variant="secondary-outline"
        color="black"
        onClick={onClose}
        width={'10em'}
        mt={'xl'}
      >
        Exit
      </Button>
    </Fragment>
  );
};

export default InProgress;
