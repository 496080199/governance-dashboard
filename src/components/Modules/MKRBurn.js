import React, { Fragment, useState } from 'react';
import styled from 'styled-components';
import {
  Box,
  Button,
  Card,
  CardBody,
  Flex,
  Grid,
  Text,
  Link
} from '@makerdao/ui-components-core';
import ModalPortal from '../ModalPortal';
import warning from '../../imgs/warning.svg';

const WarningIcon = styled.p`
  width: 63px;
  height: 57px;
  background-color: #f77249;
  mask: url(${warning}) center no-repeat;
`;

export default () => {
  const ModalTrigger = ({ text, onOpen, buttonRef }) => (
    <Button variant="danger-outline" onClick={onOpen} innerRef={buttonRef}>
      {text}
    </Button>
  );
  const Step0 = ({ onClose }) => {
    return (
      <Fragment>
        <WarningIcon />
        <Text.h2 mt="m">Are you sure you want to burn MKR?</Text.h2>
        <Text.p mt="m" mx="l" textAlign="center">
          {`By burning your MKR in the ESM, you are contributing to the
           shutdown of the Dai Credit System. Your MKR will be immediately burned
            and cannot be retrieved.`}
          {` `}
          <Link
            target="_blank"
            rel="noopener noreferrer"
            css="text-decoration: none"
          >
            Read the ESM documentation.
          </Link>
        </Text.p>
        <Flex flexDirection="row" justifyContent="space-around" m={'m'}>
          <Button
            variant="secondary-outline"
            color="black"
            onClick={onClose}
            mr={'s'}
          >
            Cancel
          </Button>
          <Button variant="danger" ml={'s'}>
            Continue
          </Button>
        </Flex>
      </Fragment>
    );
  };
  // const step1
  // const step2
  // const step3

  const ModalContent = ({ onClose }) => {
    const [step, setStep] = useState(0);
    const renderStep = step => {
      switch (step) {
        case 0:
          return <Step0 onClose={onClose} />;
        // case 1:
        //   return <Step1 />
        // case 2:
        //   return <Step2 />
        // case 3:
        //   return <Step3 />
        default:
          return <Step0 />;
      }
    };
    const renderedStep = renderStep(step);
    return (
      <Flex flexDirection="column" alignItems="center">
        {renderedStep}
      </Flex>
    );
  };

  const modalProps = {
    triggerText: 'Burn your MKR',
    ariaLabel: 'Modal to confirm burning your MKR for an emergency shutdown',
    ModalTrigger
  };
  const [mkrStaked, setMkrStaked] = useState('0.00');
  return (
    <Grid gridRowGap="m" my={'s'}>
      <Text.h4 textAlign="left" fontWeight="700">
        Total MKR Burned
      </Text.h4>
      <Card>
        <CardBody p={'s'} pb={'m'}>
          <Flex flexDirection="row" m={'s'}>
            {/* Load Number */}
            <Text.h3>
              {`${mkrStaked} MKR `}
              {` `}
            </Text.h3>
            <Text.p color="#708390" ml="xs" fontWeight="400">
              {' '}
              of 50,000 MKR
            </Text.p>
          </Flex>
          <Box
            flexGrow="1"
            bg="#F6F8F9"
            border="default"
            height="20"
            mx="s"
            my="s"
            mb="m"
            style={{ borderRadius: 5, minHeight: 20 }}
          ></Box>
        </CardBody>
        <CardBody>
          <Flex flexDirection="row" justifyContent="space-between" m={'m'}>
            <ModalPortal {...modalProps}>
              <ModalContent />
            </ModalPortal>
            <Text.p color="#9FAFB9" fontWeight="300" alignSelf="center">
              You have no MKR in the ESM
            </Text.p>
          </Flex>
        </CardBody>
      </Card>
    </Grid>
  );
};
