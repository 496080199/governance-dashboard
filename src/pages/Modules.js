import React, { useState } from 'react';
import { Flex, Grid, Text, Link } from '@makerdao/ui-components-core';

import MKRBurn from '../components/Modules/MKRBurn';
import ESMHistory from '../components/Modules/ESMHistory';

export default class ESM extends React.Component {
  state = {
    esmCliDocContent: null
  };

  componentDidMount() {
    fetch(require('../esCliDoc.md'))
      .then(res => res.text())
      .then(esmCliDocContent => this.setState({ esmCliDocContent }));
  }

  render() {
    const { esmCliDocContent } = this.state;
    return (
      <Flex flexDirection="column" minHeight="100vh">
        <Grid gridRowGap="m" mx={'2xl'} my={'2xl'} px={'2xl'}>
          <Text.h2 textAlign="left">Emergency Shutdown Module</Text.h2>
          <Text.p textAlign="justify">
            The ESM allows MKR holders to shutdown the system without a central
            authority. Once 50,000 MKR are entered into the ESM, emergency
            shutdown can be executed.
            {` `}
            <Link
              target="_blank"
              rel="noopener noreferrer"
              css="text-decoration: none"
            >
              Read the documentation here.
            </Link>
          </Text.p>
          <MKRBurn />
          <ESMHistory />
        </Grid>
      </Flex>
    );
  }
}
