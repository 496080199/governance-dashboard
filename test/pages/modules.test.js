import React from 'react';
import { Provider } from 'react-redux';
import Modules from '../../src/pages/Modules';
import { render } from '@testing-library/react';
import { takeSnapshot, restoreSnapshot } from '@makerdao/test-helpers';
import instantiateMaker from '../helpers/maker';
import Maker, { ETH, MKR } from '@makerdao/dai';
import {
  act,
  cleanup,
  fireEvent,
  wait,
  waitForElement
} from '@testing-library/react';
import governancePlugin from '@makerdao/dai-plugin-governance';
import trezorPlugin from '@makerdao/dai-plugin-trezor-web';
import ledgerPlugin from '@makerdao/dai-plugin-ledger-web';
const { click } = fireEvent;

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

let store, maker;

beforeAll(async () => {
  maker = await instantiateMaker('test');
});

afterEach(cleanup);

describe('renders summary page', () => {
  let renderedComponent;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  beforeAll(async () => {
    jest.setTimeout(20000);
    await wait(() => window.maker);
    store = mockStore({
      accounts: {
        activeAccount: maker.currentAddress(),
        allAccounts: [
          {
            address: maker.currentAddress(),
            type: 'browser',
            mkrInEsm: MKR(0),
            mkrBalance: '0.0000',
            hasProxy: false,
            proxyRole: '',
            votingFor: [],
            hasInfMkrApproval: false,
            hasInfIouApproval: false,
            proxy: {
              address: '',
              votingPower: 0,
              hasInfIouApproval: false,
              hasInfMkrApproval: false,
              linkedAccount: {}
            },
            mkrLockedChiefHot: 0,
            mkrLockedChiefCold: 0
          }
        ]
      },
      esm: {
        cageTime: 0,
        totalStaked: MKR(1),
        thresholdAmount: 50000,
        fired: false,
        canFire: false
      }
    });
  });

  test('renders with state', async () => {
    let renderedComponent = await render(<Modules store={store} />);
    expect(renderedComponent).toMatchSnapshot();
  });

  test('show progress bar', async () => {
    const { getByTestId } = await render(<Modules store={store} />);
    getByTestId('progress-bar');
  });

  test('show esm history', async () => {
    const { getByText } = await render(<Modules store={store} />);
    getByText('ESM History');
  });

  test('show "Burn your MKR" button', async () => {
    const { getByText } = await render(<Modules store={store} />);
    getByText('Burn your MKR');
  });
});
