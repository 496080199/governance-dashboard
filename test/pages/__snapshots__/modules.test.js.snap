// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders summary page renders with state 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-AykKI Sfztm sc-fzXfMA kDQaBL"
      >
        <div
          class="sc-AykKI eUTaho sc-fzXfNf hYOpSW"
        >
          <h2
            class="sc-fzXfLV sc-fzXfLW cYPrAS"
          >
            Emergency Shutdown Module
          </h2>
          <p
            class="sc-fzXfLQ HiEwu"
          >
            The ESM allows MKR holders to shutdown the system without a central authority. Once 50,000 MKR are entered into the ESM, emergency shutdown can be executed.
             
            <a
              class="sc-fzXfNl ehEzPg"
              href="https://docs.makerdao.com/smart-contract-modules/emergency-shutdown-module"
              rel="noopener noreferrer"
              target="_blank"
            >
              Read the documentation here.
            </a>
          </p>
          <div
            class="sc-AykKI bVpOEa sc-fzXfNf hYOpSW"
          >
            <h4
              class="sc-fzXfMw sc-fzXfMx ktDqRI"
              font-weight="700"
            >
              Total MKR Burned
            </h4>
            <div
              class="sc-AykKD eHBufN"
            >
              <div
                class="sc-AykKC cOjTfu"
              >
                <div
                  class="sc-AykKI kpMviG sc-fzXfMA lghjaB"
                >
                  <h3
                    class="sc-fzXfLX sc-fzXfMv kUQhnt"
                  >
                    1.00 MKR
                  </h3>
                  <p
                    class="sc-fzXfLQ gDZiLf"
                    color="#708390"
                    font-weight="400"
                  >
                     
                    of 
                    50000
                  </p>
                </div>
                <div
                  class="sc-AykKI DMwiT"
                  data-testid="progress-bar"
                  height="20"
                  style="border-radius: 5px; min-height: 20px;"
                >
                  <div
                    class="sc-fzXfPF bBgSKP"
                    style="width: 0%;"
                  />
                </div>
              </div>
              <div
                class="sc-AykKC QXPLA"
              >
                <div
                  class="sc-AykKI jmfkqQ sc-fzXfMA heGUAX"
                >
                  <button
                    class="sc-fzXfLO ipwkKa"
                  >
                    Burn your MKR
                  </button>
                  <p
                    class="sc-fzXfLQ OvbJe"
                    color="#9FAFB9"
                    font-weight="300"
                  >
                    You have no MKR in the ESM
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-AykKI bVpOEa sc-fzXfNf hYOpSW"
          >
            <h4
              class="sc-fzXfMw sc-fzXfMx ktDqRI"
              font-weight="700"
            >
              ESM History
            </h4>
            <div
              class="sc-AykKD gIqJED"
            >
              <table
                class="sc-fzXfNS jeSlkx normal"
                width="100%"
              >
                <thead>
                  <tr>
                    <th>
                      Date
                    </th>
                    <th>
                      Amount Staked
                    </th>
                    <th>
                      Sender Address
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td
                      colspan="3"
                    >
                      <span
                        class="sc-fzXfLP cjzehg"
                        color="darkLavender"
                      >
                        Loading
                      </span>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-AykKI Sfztm sc-fzXfMA kDQaBL"
    >
      <div
        class="sc-AykKI eUTaho sc-fzXfNf hYOpSW"
      >
        <h2
          class="sc-fzXfLV sc-fzXfLW cYPrAS"
        >
          Emergency Shutdown Module
        </h2>
        <p
          class="sc-fzXfLQ HiEwu"
        >
          The ESM allows MKR holders to shutdown the system without a central authority. Once 50,000 MKR are entered into the ESM, emergency shutdown can be executed.
           
          <a
            class="sc-fzXfNl ehEzPg"
            href="https://docs.makerdao.com/smart-contract-modules/emergency-shutdown-module"
            rel="noopener noreferrer"
            target="_blank"
          >
            Read the documentation here.
          </a>
        </p>
        <div
          class="sc-AykKI bVpOEa sc-fzXfNf hYOpSW"
        >
          <h4
            class="sc-fzXfMw sc-fzXfMx ktDqRI"
            font-weight="700"
          >
            Total MKR Burned
          </h4>
          <div
            class="sc-AykKD eHBufN"
          >
            <div
              class="sc-AykKC cOjTfu"
            >
              <div
                class="sc-AykKI kpMviG sc-fzXfMA lghjaB"
              >
                <h3
                  class="sc-fzXfLX sc-fzXfMv kUQhnt"
                >
                  1.00 MKR
                </h3>
                <p
                  class="sc-fzXfLQ gDZiLf"
                  color="#708390"
                  font-weight="400"
                >
                   
                  of 
                  50000
                </p>
              </div>
              <div
                class="sc-AykKI DMwiT"
                data-testid="progress-bar"
                height="20"
                style="border-radius: 5px; min-height: 20px;"
              >
                <div
                  class="sc-fzXfPF bBgSKP"
                  style="width: 0%;"
                />
              </div>
            </div>
            <div
              class="sc-AykKC QXPLA"
            >
              <div
                class="sc-AykKI jmfkqQ sc-fzXfMA heGUAX"
              >
                <button
                  class="sc-fzXfLO ipwkKa"
                >
                  Burn your MKR
                </button>
                <p
                  class="sc-fzXfLQ OvbJe"
                  color="#9FAFB9"
                  font-weight="300"
                >
                  You have no MKR in the ESM
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-AykKI bVpOEa sc-fzXfNf hYOpSW"
        >
          <h4
            class="sc-fzXfMw sc-fzXfMx ktDqRI"
            font-weight="700"
          >
            ESM History
          </h4>
          <div
            class="sc-AykKD gIqJED"
          >
            <table
              class="sc-fzXfNS jeSlkx normal"
              width="100%"
            >
              <thead>
                <tr>
                  <th>
                    Date
                  </th>
                  <th>
                    Amount Staked
                  </th>
                  <th>
                    Sender Address
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td
                    colspan="3"
                  >
                    <span
                      class="sc-fzXfLP cjzehg"
                      color="darkLavender"
                    >
                      Loading
                    </span>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
